rust/src/dns/dns.rs

```diff
@@ -0,0 +1,801 @@
+extern crate nom;
+
+use std;
+use std::mem::transmute;
+
+use log::*;
+use core;
+use dns::parser;
+
+/// DNS record types.
+pub const DNS_RTYPE_A:     u16 = 1;
+pub const DNS_RTYPE_CNAME: u16 = 5;
+pub const DNS_RTYPE_SOA:   u16 = 6;
+pub const DNS_RTYPE_PTR:   u16 = 12;
+pub const DNS_RTYPE_MX:    u16 = 15;
+pub const DNS_RTYPE_AAAA:  u16 = 28;
+pub const DNS_RTYPE_SSHFP: u16 = 44;
+pub const DNS_RTYPE_RRSIG: u16 = 46;
+
+/// DNS error codes.
+pub const DNS_RCODE_NOERROR:  u16 = 0;
+pub const DNS_RCODE_FORMERR:  u16 = 1;
+pub const DNS_RCODE_NXDOMAIN: u16 = 3;
+
+/// The maximum number of transactions to keep in the queue pending
+/// processing before they are aggressively purged. Due to the
+/// stateless nature of this parser this is rarely needed, especially
+/// when one call to parse a request parses and a single request, and
+/// likewise for responses.
+///
+/// Where this matters is when one TCP buffer contains multiple
+/// requests are responses and one call into the parser creates
+/// multiple transactions. In this case we have to hold onto
+/// transactions longer than until handling the next transaction so it
+/// gets logged.
+const MAX_TRANSACTIONS: usize = 32;
+
+#[repr(u32)]
+pub enum DNSEvent {
+    UnsolicitedResponse = 0,
+    MalformedData,
+    NotRequest,
+    NotResponse,
+    ZFlagSet,
+    Flooded,
+    StateMemCapReached,
+}
+
+#[derive(Debug,PartialEq)]
+pub struct DNSHeader {
+    pub tx_id: u16,
+    pub flags: u16,
+    pub questions: u16,
+    pub answer_rr: u16,
+    pub authority_rr: u16,
+    pub additional_rr: u16,
+}
+
+#[derive(Debug)]
+pub struct DNSQueryEntry {
+    pub name: Vec<u8>,
+    pub rrtype: u16,
+    pub rrclass: u16,
+}
+
+impl DNSQueryEntry {
+
+    pub fn name(&self) -> &str {
+        let r = std::str::from_utf8(&self.name);
+        if r.is_err() {
+            return "";
+        }
+        return r.unwrap();
+    }
+
+}
+
+#[derive(Debug,PartialEq)]
+pub struct DNSAnswerEntry {
+    pub name: Vec<u8>,
+    pub rrtype: u16,
+    pub rrclass: u16,
+    pub ttl: u32,
+    pub data_len: u16,
+    pub data: Vec<u8>,
+}
+
+impl DNSAnswerEntry {
+
+    pub fn name(&self) -> &str {
+        let r = std::str::from_utf8(&self.name);
+        if r.is_err() {
+            return "";
+        }
+        return r.unwrap();
+    }
+
+    pub fn data_to_string(&self) -> &str {
+        let r = std::str::from_utf8(&self.data);
+        if r.is_err() {
+            return "";
+        }
+        return r.unwrap();
+    }
+
+}
+
+#[derive(Debug)]
+pub struct DNSRequest {
+    pub header: DNSHeader,
+    pub queries: Vec<DNSQueryEntry>,
+}
+
+#[derive(Debug)]
+pub struct DNSResponse {
+    pub header: DNSHeader,
+    pub queries: Vec<DNSQueryEntry>,
+    pub answers: Vec<DNSAnswerEntry>,
+    pub authorities: Vec<DNSAnswerEntry>,
+}
+
+#[derive(Debug)]
+pub struct DNSTransaction {
+    pub id: u64,
+    pub request: Option<DNSRequest>,
+    pub response: Option<DNSResponse>,
+    pub logged: u32,
```

Comment:
would it make sense to define our own type in rust so that if we change things on the C side we can easily update it for all Rust parsers? Same would apply to C itself btw.